### PR TITLE
Improve exception grouping on Android for RuntimeExceptions

### PR DESCRIPTION
**Pull request by markushi**
## :scroll: Description
Add `exception_id` and `parent_id` to describe the outer/inner exception relationships. 

Also add `is_exception_group=true` to `java.lang.RuntimeException` - not sure if we should apply this globally or not? Maybe @adinauer can provide some backend feedback here?

## :bulb: Motivation and Context
Fixes https://github.com/getsentry/sentry/issues/64074
This affects exception grouping and determining issue titles, [RFC details are here](https://github.com/getsentry/rfcs/blob/main/text/0079-exception-groups.md#sentry-issue-grouping).

Funny enough the same outcome can be achieved by doing nothing from the above and simply sending the exception->values in reverse order ([right now it's](https://github.com/getsentry/sentry-java/blob/5b57ad1e54e8ab749aabb392f4ce8d6f116215b8/sentry/src/main/java/io/sentry/SentryExceptionFactory.java#L187) `values = [inner, outer]`)

## :pencil: Checklist

- [x] I reviewed the submitted code.
- [x] I added tests to verify the changes.
- [x] No new PII added or SDK only sends newly added PII if `sendDefaultPII` is enabled.
- [x] I updated the docs if needed.
- [x] Review from the native team if needed.
- [x] No breaking change or entry added to the changelog.
- [x] No breaking change for hybrid SDKs or communicated to hybrid SDKs.

### CODE REVIEW

File: sentry/src/main/java/io/sentry/UncaughtExceptionHandlerIntegration.java [hunk 1]
```diff
@@ -147,6 +147,11 @@ static Throwable getUnhandledThrowable(
     final Mechanism mechanism = new Mechanism();
     mechanism.setHandled(false);
     mechanism.setType("UncaughtExceptionHandler");
+
+    // probably should be Android only?
+    if (thrown instanceof RuntimeException) {
```
Comment (adinauer) [inline sentry/src/main/java/io/sentry/UncaughtExceptionHandlerIntegration.java:152]: A `RuntimeException` by itself isn't an exception group, there's suppressed exceptions which we have an issue for (https://github.com/getsentry/sentry-java/issues/542) but I'm guessing that has little to do with the issue we're trying to fix here.

**Comment by adinauer** [comment]
> Funny enough the same outcome can be achieved by doing nothing from the above and simply sending the exception->values in reverse order ([right now it's](https://github.com/getsentry/sentry-java/blob/5b57ad1e54e8ab749aabb392f4ce8d6f116215b8/sentry/src/main/java/io/sentry/SentryExceptionFactory.java#L187) values = [inner, outer])

This sounds like something that wouldn't apply to all examples. Let's assume a `NullPointerException` which is caught and rethrown a bunch of times. Now multiple separate code problems would end up in the same NPE issue, just because we use the innermost exception. They should still end up as separate issues in Sentry though.

My suggestion would be to unwrap the `RuntimeException` in case of Android if we're sure this is causing problems.

I'm not a fan of hacking this in via the exception groups feature as I'm unsure whether this would cause problems when we actually get around to implementing suppressed exception support (https://github.com/getsentry/sentry-java/issues/542).

**Comment by lobsterkatie** [comment]
Yeah, we definitely don't want to just reverse the order, for exactly the reasons @adinauer said, and because it would confuse the UI.

Also, IMHO it should definitely be scoped to this particular wrapper exception, lest we overcompensate and end up creating the exact problem that JS is having in https://github.com/getsentry/sentry/issues/59679.

As for unwrapping vs using `is_exception_group`:

- Are the error message and stacktrace of the wrapper error always the same and/or do they provide any informational value? If no, then I'd vote for unwrapping.
- I agree that this is kind of abuse of `is_exception_group`, in order to trick our logic [here](https://github.com/getsentry/sentry/blob/f8ad4ec1d5b90e74381ca50987dc436fb5c8cbf1/src/sentry/grouping/strategies/newstyle.py#L770-L782) and [here](https://github.com/getsentry/sentry/blob/f8ad4ec1d5b90e74381ca50987dc436fb5c8cbf1/src/sentry/grouping/strategies/newstyle.py#L809C5-L818) into setting the correct `main_exception_id`. It would be a stopgap until the time if and when we make this logic smarter. See https://github.com/getsentry/sentry/issues/64088.

**Comment by romtsn** [comment]
> Are the error message and stacktrace of the wrapper error always the same and/or do they provide any informational value? If no, then I'd vote for unwrapping.

Most of the time should be the same, but might differ between OS versions depending on what they've changed. I agree we should scope this to Android only and probably unwrapping is a better way forward here. This is also only a case for checked exceptions, so if we could narrow down to only those on Android would be great.

**Comment by adinauer** [comment]
Could we do the unwrapping during Sentry processing instead of the SDK?
- Not sure if obfuscation is a problem here
- We could adapt this for future (Android) versions without requiring customers to upgrade their SDK and their users in turn to upgrade the App.

**Comment by markushi** [comment]
Thanks for chiming in @lobsterkatie!

> Also, IMHO it should definitely be scoped to this particular wrapper exception, lest we overcompensate and end up creating the exact problem that JS is having in https://github.com/getsentry/sentry/issues/59679.

Agreed! Unfortunately we could only do a best-effort here, as this `java.lang.RuntimeException` is not specific to Android, and could be also thrown by anyone anywhere.

> Are the error message and stacktrace of the wrapper error always the same and/or do they provide any informational value? 

Whilst the wrapped exception itself and it's stacktrace don't provide any extra information, it indicates one crucial semantic information: An app crash is about to happen.

Thinking more about future solutions: I'm wondering if in situations where the wrapping exception contains no in-app frames, we simply could decide to just use the inner exception for titleing.

**Comment by lobsterkatie** [comment]
> I'm wondering if in situations where the wrapping exception contains no in-app frames, we simply could decide to just use the inner exception for titleing.

Interesting. I added this in https://github.com/getsentry/sentry/issues/64088 as an idea to consider.

**Comment by adinauer** [comment]
> I'm wondering if in situations where the wrapping exception contains no in-app frames, we simply could decide to just use the inner exception for titleing.

Hmm haven't thought this through but e.g. config issues could end up with no in-app frames but still be caused by the user. Same goes for library bugs that may e.g. affect request handling where the exception doesn't go through user code.

**Comment by markushi** [comment]
In the meantime I stumbled upon the synthetic flag in our [exception interface](https://develop.sentry.dev/sdk/event-payloads/exception/#exception-mechanism).

> An optional flag indicating that this error is synthetic. Synthetic errors are errors that carry little meaning by themselves. This may be because they are created at a central place (like a crash handler), and are all called the same: Error, Segfault etc. When the flag is set, Sentry will then try to use other information (top in-app frame function) rather than exception type and value in the UI for the primary event display. 

IMHO this sounds awfully similar to what we actually want to achieve here. @lobsterkatie would this make sense for you?

**Comment by markushi** [comment]
Closing this one for now in favor of the different issues linked above.

**Pull request closed without merging**